lib/workers/repository/process/vulnerabilities.ts

```diff
@@ -91,51 +97,367 @@ export class Vulnerabilities {
+      packageName = packageName.toLowerCase().replace(regEx(/[_.-]+/g), '-');
+    }
+
+    const packageRules: PackageRule[] = [];
+    try {
+      const vulnerabilities = await this.osvOffline?.getVulnerabilities(
+        ecosystem,
+        packageName
+      );
+      if (
+        is.nullOrUndefined(vulnerabilities) ||
+        is.emptyArray(vulnerabilities)
+      ) {
+        logger.trace(
+          `No vulnerabilities found in OSV database for ${packageName}`
+        );
+        return [];
+      }
+
+      const depVersion =
+        dep.lockedVersion ?? dep.currentVersion ?? dep.currentValue!;
+
+      const versioning = dep.versioning ?? getDefaultVersioning(dep.datasource);
+      const versioningApi = getVersioning(versioning);
 
-    const vulnerabilities = await this.osvOffline?.getVulnerabilities(
-      ecosystem!,
-      packageDependency.depName!
+      if (!versioningApi.isVersion(depVersion)) {
+        logger.debug(
+          `Skipping vulnerability lookup for package ${packageName} due to unsupported version ${depVersion}`
+        );
+        return [];
+      }
+
+      for (const vulnerability of vulnerabilities) {
+        for (const affected of vulnerability.affected ?? []) {
+          if (
+            this.isPackageVulnerable(
+              ecosystem,
+              packageName,
+              depVersion,
+              affected,
+              versioningApi
+            )
+          ) {
+            logger.debug(
+              `Vulnerability ${vulnerability.id} affects ${packageName} ${depVersion}`
+            );
+            const fixedVersion = this.getFixedVersion(
+              ecosystem,
+              depVersion,
+              affected,
+              versioningApi
+            );
+            if (fixedVersion) {
+              logger.debug(
+                `Setting allowed version ${fixedVersion} to fix vulnerability ${vulnerability.id} in ${packageName} ${depVersion}`
+              );
+              const rule = this.convertToPackageRule(
+                packageFileConfig,
+                dep,
+                packageName,
+                depVersion,
+                fixedVersion,
+                vulnerability
+              );
+              packageRules.push(rule);
+            } else {
+              logger.debug(
+                `No fixed version available for vulnerability ${vulnerability.id} in ${packageName} ${depVersion}`
+              );
+            }
+          }
+        }
+      }
+
+      this.sortByFixedVersion(packageRules, versioningApi);
+    } catch (err) {
+      logger.debug(
+        { err },
+        `Error fetching vulnerability information for ${packageName}`
+      );
+    }
+
+    return packageRules;
+  }
+
+  private sortByFixedVersion(
+    packageRules: PackageRule[],
+    versioningApi: VersioningApi
+  ): void {
+    packageRules.sort((a, b) =>
+      versioningApi.sortVersions(
+        (a.allowedVersions as string).replace(regEx(/[=> ]+/g), ''),
```

Comment:
this is heavy, please create a version map to speed up sorting or preprocess to separate field before sorting